sqlmodel/_compat.py

```diff
@@ -346,6 +346,16 @@ def sqlmodel_init(*, self: "SQLModel", data: Dict[str, Any]) -> None:
                 self_instance=self,
             )
         else:
+            raw_self = self.model_copy()
+            pydantic_validated_model = self.__pydantic_validator__.validate_python(
+                data,
+                self_instance=raw_self,
```

Comment:
why was `raw_self` necessary here?

Reply:
> why was `raw_self` necessary here?

```python
self.__pydantic_validator__.validate_python(
                data,
                self_instance=raw_self,
            )
```
would update self, and something would goes wrong on `sqlmodel_table_construct` for the changed self. Maybe that's the reason that the original code dosen't validate when `table=True`.

Actually, we need to get the pydantic-validated dict, and use this dict to update the original dict, and then construct tables by the updated-original dict.

 Anyway, you can try `self_instance=self`, and some unittests would fail.